utensor_cgen/backend/graph_lower/alloc_plan.py

```diff
@@ -0,0 +1,112 @@
+import attr
+import six
+from attr.validators import instance_of
+
+from utensor_cgen.logger import logger
+
+__all__ = ['TimeslotAllocation', 'SpaceAllocation', 'TimeSpaceAllocation', 'AllocationPlan']
+
+@attr.s
+class TimeslotAllocation(object):
+  time_slot_start = attr.ib(validator=instance_of(int))
+  # if time_slot_end is None, it's a time sapn with no end
+  time_slot_end = attr.ib(validator=instance_of((int, type(None))))
+
+  def __attrs_post_init__(self):
+    assert self.time_slot_start >= 0, \
+      'invalid time_slot_start: %s' % self.time_slot_start
+    if self.time_slot_end is not None:
+      assert self.time_slot_end >= self.time_slot_start, \
+        'invalid time_slot_end: %s ~ %s' % (self.time_slot_start, self.time_slot_end)
+  
+  def __contains__(self, slot):
+    assert isinstance(slot, int), 'incorrect slot type: %s' % type(slot)
+    is_in = self.time_slot_start <= slot
+    if self.time_slot_end is not None:
+      is_in = is_in and slot <= self.time_slot_end
+    return is_in
+
+
+@attr.s
+class SpaceAllocation(object):
```

Comment:
This will need to be able to account for memory access alignment. Probably need to include this info in the attributes

Reply:
I add one more config field, `data_alignment`, for the `TensorAllocationPlanner`.
Now the planner is aware of data alignment and it's configurable to account for different platform.
It's default to 2-bytes alignment.
I'll add this attribute to `SpaceAllocation`.

Reply:
done

Reply:
marvelous :)